File: cmd/tui/nfo.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func writeTvShowNFO(dir, title, plot string, tmdbID int) error {
	var b strings.Builder
	b.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<tvshow>\n")
	b.WriteString(xmlTag("title", title))
	if plot != "" {
		b.WriteString(xmlTag("plot", plot))
	}
	if tmdbID > 0 {
		b.WriteString(fmt.Sprintf("  <uniqueid type=\"tmdb\" default=\"true\">%d</uniqueid>\n", tmdbID))
	}
	b.WriteString("</tvshow>\n")
	return os.WriteFile(filepath.Join(dir, "tvshow.nfo"), []byte(b.String()), 0o644)
}

func writeEpisodeNFO(mediaPath, title string, season, ep int, plot string) error {
	var b strings.Builder
	b.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<episodedetails>\n")
	b.WriteString(xmlTag("title", title))
	if season > 0 {
		b.WriteString(xmlTag("season", fmt.Sprintf("%d", season)))
	}
	if ep > 0 {
		b.WriteString(xmlTag("episode", fmt.Sprintf("%d", ep)))
	}
	if plot != "" {
		b.WriteString(xmlTag("plot", plot))
	}
	b.WriteString("</episodedetails>\n")
	nfo := strings.TrimSuffix(mediaPath, filepath.Ext(mediaPath)) + ".nfo"
	return os.WriteFile(nfo, []byte(b.String()), 0o644)
}

func xmlTag(name, v string) string {
	v = strings.ReplaceAll(v, "&", "&amp;")
	v = strings.ReplaceAll(v, "<", "&lt;")
	v = strings.ReplaceAll(v, ">", "&gt;")
	return fmt.Sprintf("  <%s>%s</%s>\n", name, v, name)
}

File: src/lib/nfo.ts
import { writeFileSync } from 'node:fs'
import { extname, join } from 'node:path'

function xmlTag(name: string, v: string): string {
  const safe = v.replaceAll('&', '&amp;').replaceAll('<', '&lt;').replaceAll('>', '&gt;')
  return `  <${name}>${safe}</${name}>\n`
}

export function writeTvShowNFO(dir: string, title: string, plot: string, tmdbId: number): void {
  let b = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<tvshow>\n'
  b += xmlTag('title', title)
  if (plot) b += xmlTag('plot', plot)
  if (tmdbId > 0) b += `  <uniqueid type="tmdb" default="true">${tmdbId}</uniqueid>\n`
  b += '</tvshow>\n'
  writeFileSync(join(dir, 'tvshow.nfo'), b)
}

export function writeEpisodeNFO(mediaPath: string, title: string, season: number, ep: number, plot: string): void {
  let b = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n<episodedetails>\n'
  b += xmlTag('title', title)
  if (season > 0) b += xmlTag('season', String(season))
  if (ep > 0) b += xmlTag('episode', String(ep))
  if (plot) b += xmlTag('plot', plot)
  b += '</episodedetails>\n'
  writeFileSync(mediaPath.slice(0, mediaPath.length - extname(mediaPath).length) + '.nfo', b)
}
